src/sidebar/oauth-auth.js

```diff
@@ -223,17 +215,25 @@ function auth($http, $rootScope, $window,
    */
   function tokenGetter() {
     if (!tokenInfoPromise) {
-      var grantToken = grantTokenFromHostPage();
-
-      if (grantToken) {
-        // Exchange host-page provided grant token for a new access token.
-        tokenInfoPromise = exchangeJWT(grantToken).then((tokenInfo) => {
-          return tokenInfo;
-        }).catch(function(err) {
-          showAccessTokenExpiredErrorMessage(
-            'You must reload the page to annotate.');
-          throw err;
-        });
+      var cfg = serviceConfig(settings);
+
+      // Check if automatic login is being used, indicated by the presence of
+      // the 'grantToken' property in the service configuration.
+      if (cfg && typeof cfg.grantToken !== 'undefined') {
+        if (cfg.grantToken) {
+          // User is logged-in on the publisher's website.
+          // Exchange the grant token for a new access token.
+          tokenInfoPromise = exchangeJWT(cfg.grantToken).then((tokenInfo) => {
+            return tokenInfo;
+          }).catch(function(err) {
+            showAccessTokenExpiredErrorMessage(
+              'You must reload the page to annotate.');
+            throw err;
+          });
+        } else {
+          // User is anonymous on the publisher's website.
+          tokenInfoPromise = Promise.resolve(null);
+        }
```

Comment:
The logic here, if I understand it correctly, is that whenever there's a truthy `serviceConfig()` (which currently I believe is equivalent to whenever the embedding page contains a `js-hypothesis-config` or a `window.hypothesisConfig()` with a `services` object) then first-party accounts are disabled and third-party accounts are used.

But in the longer term I think what we want is for the default hypothes.is service to be a service like any other, one that appears in the `services` object alongside any other service configs, it's just that it's found in a `services` object in the `app.html` page instead of in the embedding page.

Do you agree? And if so, do we need another way of deciding when to enable third-party accounts, instead of checking for the presence of a service config?

I guess this also introduces another way to disable Hypothesis on a site, by embedding the client and including a services config (e.g. one with an authority that doesn't exist server-side and a grant token that doesn't work).

Reply:
> The logic here, if I understand it correctly, is that whenever there's a truthy serviceConfig() (which currently I believe is equivalent to whenever the embedding page contains a js-hypothesis-config or a window.hypothesisConfig() with a services object) then first-party accounts are disabled and third-party accounts are used.

Here we aren't testing for first vs third party services but rather the use of automatic login via a grant token. That is what matters because this method is responsible for obtaining an access token from somewhere. In future if h becomes "just another annotation service" then it would, in theory, be eligible to use grant tokens for login as well.

> But in the longer term I think what we want is for the default hypothes.is service to be a service like any other, one that appears in the services object alongside any other service configs, it's just that it's found in a services object in the app.html page instead of in the embedding page.

Yes, that is where I expect we'll eventually end up.